### DISCUSSION TITLE
Is it possible to change the color in stamp_text?

### DISCUSSION THREAD

**Discussion by Vishwamithra37** [MatthiasValvekens/pyHanko#107] [Q&A] [answered]
I want to cchange the color inside stamp_text for each line (for every \n) and just want to add my own style to the contents of the text. 
How do I go about this using pyHanko inbuilt textbox?

**Comment by MatthiasValvekens** [comment, marked as answer]
Unfortunately not natively, no. While changing text colours is something that would be (relatively) easy to implement---modulo subtleties with colour spaces---implementing even "basic" styling is a lot less straightforward than it sounds. Rich text layout in PDF requires the PDF creation software to juggle multiple font resources, deal with paragraph layout, text shaping, etc. The general problem is a lot harder than in .docx/.odt/.html/.epub/....


Right now, pyHanko's appearance generation code only performs very primitive text layout: no word wrapping, use one single font, no frills. On the other hand, I did implement OTF font support with shaping performed through HarfBuzz. That said, the main reason for doing so was enabling support for CJK scripts and embedding font subsets, not fancy styling.

Automatic paragraph layout is a tricky problem that's way beyond pyHanko's scope (it would add a lot of potentially hard-to-maintain code that has nothing to do with digital signing, basically). I am however willing to accept PRs for the following features:

 - Coloured text (again, this one's fairly easy)
 - Multi-font support (this one requires fairly deep knowledge of the PDF graphics model and font resource handling)
 - Basic rich text, i.e. bold/italic/underline (requires multifont support!) and perhaps simple word wrapping based on character count.

If anyone out there wants to work on one or more of these: please open a discussion so we can discuss implementation strategies.


--------------------

So, what options do you have in the meantime?

-  Generate an appearance using external tools, save it to PDF and import the resulting graphics wholesale using `StaticStampStyle`.
- Hack together your own `TextStampStyle` subclass that works for your use case. Changing the colour (assuming RGB) is done using the `rg` operator.

**Reply by Vishwamithra37** [reply]
If my application using pyHanko starts generating money, I'd definetely invest on those tasks :).
